inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/PdfAnnoRenderer.java

```diff
@@ -108,35 +110,62 @@ public static PdfAnnoModel render(AnnotatorState aState, VDocument aVDoc, String
-            PdfExtractLine last = aPdfExtractFile.getStringPdfExtractLine(end);
-            return new Span(aVspan.getVid().getId(), first.getPage(), aColor,
-                cleanAnnotatedText.substring(start - index, end - index).replaceAll("\\s", ""),
-                first.getPosition(), last.getPosition());
+        int docTextLen = aDocumentText.length();
+        int vSpanBegin = aDocumentText.length();
+        int vSpanEnd = -1;
+        // search for begin of the first range and end of the last range and use those
+        for (VRange range : aVspan.getRanges())
+        {
+            if (range.getBegin() < vSpanBegin) vSpanBegin = range.getBegin();
+            if (range.getEnd() > vSpanEnd) vSpanEnd = range.getEnd();
         }
+
+        // use an offset before and after string to find a unique text snippet for an annotation
+        // begin with 0 and increase by 1 until a unique text is found
+        // offset is limited to length of a quarter of the text
+        for (int offset = 0; offset < docTextLen / 4 + 1; offset++)
+        {
+            // subtract offset from vSpanBegin and add offset to vSpanEnd and stay in bounds
+            int offsetBegin = vSpanBegin <= offset ? 0 : vSpanBegin - offset;
+            int offsetEnd = vSpanEnd < docTextLen - offset ? vSpanEnd + offset : docTextLen - 1;
+
+            // get annotated text from aDocumentText with offset before and after it
+            // also remove all whitespaces because they do not exist in PDFExtract text
+            String annotatedText = aDocumentText.substring(offsetBegin, offsetEnd).replaceAll("\\s", "");
+            String offsetBeforeText = aDocumentText.substring(offsetBegin, vSpanBegin).replaceAll("\\s", "");
+            String offsetAfterText = aDocumentText.substring(vSpanEnd, offsetEnd).replaceAll("\\s", "");
+
+            List<Integer> indices =
+                getAllIndices(aPdfExtractFile.getStringContent(), annotatedText);
+            if (indices.size() == 0) {
+                // if there are no matches end search as adding more characters won't help anymore
+                break;
+            } else if (indices.size() == 1) {
+                int index = indices.get(0);
+                // get begin and end position of the original annotationText within PDFExtract text
+                int annotationBegin = index + offsetBeforeText.length();
+                int annotationEnd = index + annotatedText.length() - offsetAfterText.length() - 1;
+                // get according PDFExtract file lines for begin and end of annotation
+                PdfExtractLine firstLine = aPdfExtractFile.getStringPdfExtractLine(annotationBegin);
+                PdfExtractLine lastLine = aPdfExtractFile.getStringPdfExtractLine(annotationEnd);
+                return new Span(aVspan.getVid().getId(), firstLine.getPage(), aColor,
+                    annotatedText.substring(annotationBegin - index, annotationEnd - index + 1),
```

Comment:
The annotatedText doesn't contain any spaces anymore, right? What is the excerpt that you create from the annotatedText here used for?

Reply:
Bad variable naming of mine... :/
Actually the variable `annotatedText` also already contains the offset text, which then needs to be removed.
I will change it to only contain the annotatedText.

Reply:
Ok, but what is it used for? I mean, the text doesn't contain any spaces... what is it good for without the spaces?

Reply:
It is for producing the Anno file for PDFAnno and the format for this file is like

```
[[spans]]
id = "1"
page = 1
label = "label"
text = "someannotatedtext"
textrange = [22,38]
```
It doesn't contain spaces, as it uses PDFExtract output without spaces.
It seems PDFAnno is not aware of spaces at all, and it does not need them because it knows the position of the characters from the PDFExtract output.

Reply:
So does it use the text? Since it has the range, the text seems redundant, doesn't it?

Reply:
I also wondered why the text is included there.
But I think it is useful when exporting annotation froms PDFAnno.
In our case it is useless I guess.
I didn't check, but at some points while developing it seemed that the content of `text` field is not relevant to PDFAnno and it could contain wrong data but still find the correct annotation position.
I am not sure if it is required at all and would cause errors if the field is completely missing.

Reply:
@reckart Should I check out if PDFAnno can import the anno file without `text` fields or can I resolve this conversation?
Even if it would work without it I'd rather keep it in because it is the actual format.

Reply:
I don't have a strong opinion on it. If you are curious, test it - otherwise resolve.

Reply:
Seems to work without the text, however when pressing export button the text is missing.
I will leave it in for now but keep it in mind.
Maybe this can be useful for performance improvement or other stuf later on if required.

Reply:
Add a comment to the source code explaining why this is added / what happens if it is omitted.

Reply:
Added it to the Span class, containing the `text` field and also producing the anno file span string.